src/main/java/bio/terra/service/snapshot/flight/create/SnapshotCreateFlight.java

```diff
@@ -430,11 +430,15 @@ public SnapshotCreateFlight(FlightMap inputParameters, Object applicationContext
             IamResourceType.DATASET,
             "A snapshot was created from this dataset."));
 
-    // at end of flight, add created snapshot id to the snapshot request
     if (mode == SnapshotRequestContentsModel.ModeEnum.BYREQUESTID) {
+      UUID snapshotRequestId = contents.getRequestIdSpec().getSnapshotRequestId();
+      // Add created snapshot id to the snapshot request.
       addStep(
           new AddCreatedSnapshotIdToSnapshotRequestStep(
-              snapshotRequestDao, contents.getRequestIdSpec().getSnapshotRequestId(), snapshotId));
+              snapshotRequestDao, snapshotRequestId, snapshotId));
+      // Notify user that snapshot is ready to use.
+      addStep(
+          new NotifyUserOfSnapshotCreationStep(snapshotBuilderService, userReq, snapshotRequestId));
```

Comment:
Just to clarify, is this the point in the code that triggers the sending of the notification that indicates that the snapshot has been created? Is the NotifyUserOfSnapshotCreationStep object what triggers the execution of the other added notification methods (if so, how)?

Reply:
Yes, this is how steps in a flight are configured. The steps themselves aren't run directly by TDR, but are run by the stairway library. It has a scheduled task which polls the database for tasks that need to be run. If it has room in its thread pool, and a flight is ready to be run, it will read it from the database and then execute the steps in order, writing a step's status to the database once it's complete. Because it persists state between steps, a flight can be interrupted at any point and then resumed from the last incomplete step.

Initially when a flight is constructed, nothing is run; instead stairway writes the information about the flight to its database so it can be later picked up.